core/src/main/scala/org/ensime/indexer/graph/GraphService.scala

```diff
@@ -26,7 +27,63 @@ import org.ensime.util.file._
 import org.ensime.vfs._
 import shapeless.cachedImplicit
 
-// FIXME: these domain objects are remnants of the DatabaseService and should be remodelled
+sealed trait FqnSymbol {
+  def fqn: String
+  def line: Option[Int]
+  def source: Option[String]
+  def declAs: DeclaredAs
+  def access: Access
+
+  def sourceFileObject(implicit vfs: EnsimeVFS): Option[FileObject] = source.map(vfs.vfile)
+}
+
+sealed trait Hierarchy {
+  def toSet: Set[ClassDef] = this match {
```

Comment:
the definition of `toSet` tells me that it is perhaps best as a typeclass or in a utility method somewhere else?